File: android/app/src/main/java/com/twintipsolutions/aclrehabtracker/util/DateHelpers.kt
package com.twintipsolutions.aclrehabtracker.util

import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.concurrent.TimeUnit

object DateHelpers {
    fun calculateWeekPostOp(surgeryDate: Date?): Int {
        if (surgeryDate == null) return 0
        val diffMillis = Date().time - surgeryDate.time
        val days = TimeUnit.MILLISECONDS.toDays(diffMillis)
        return (days / 7).toInt().coerceAtLeast(0)
    }

    fun daysUntilSurgery(surgeryDate: Date?): Int {
        if (surgeryDate == null) return 0
        val today = Calendar.getInstance().apply {
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }.time
        val surgery = Calendar.getInstance().apply {
            time = surgeryDate
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }.time
        val diffMillis = surgery.time - today.time
        return TimeUnit.MILLISECONDS.toDays(diffMillis).toInt()
    }

    fun formatDate(date: Date): String {
        return SimpleDateFormat("MMMM d, yyyy", Locale.getDefault()).format(date)
    }

    fun formatShortDate(date: Date): String {
        return SimpleDateFormat("MMM d", Locale.getDefault()).format(date)
    }

    fun formatTime(date: Date): String {
        return SimpleDateFormat("h:mm a", Locale.getDefault()).format(date)
    }

    fun formatFullDate(date: Date): String {
        return SimpleDateFormat("EEEE, MMMM d, yyyy", Locale.getDefault()).format(date)
    }

    fun todayFormatted(): String {
        return SimpleDateFormat("MMMM d", Locale.getDefault()).format(Date())
    }

    fun dateKey(date: Date): String {
        return SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(date)
    }

    fun <T> groupByDate(
        items: List<T>,
        dateExtractor: (T) -> Date
    ): List<Pair<String, List<T>>> {
        return items
            .groupBy { dateKey(dateExtractor(it)) }
            .toSortedMap(compareByDescending { it })
            .map { (key, values) ->
                val displayDate = formatDate(dateExtractor(values.first()))
                displayDate to values
            }
    }
}

File: ios/ios/ACLRehabTracker/Utilities/DateHelpers.swift
import Foundation

struct DateHelpers {
    /// Calculate the number of weeks post-operation from the surgery date
    static func calculateWeekPostOp(from surgeryDate: Date) -> Int {
        let now = Date()
        let diffTime = abs(now.timeIntervalSince(surgeryDate))
        let diffDays = Int(diffTime / (60 * 60 * 24))
        return diffDays / 7
    }

    /// Format a date for display (e.g., "January 15, 2024")
    static func formatDate(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateStyle = .long
        formatter.timeStyle = .none
        return formatter.string(from: date)
    }

    /// Format a date for short display (e.g., "Jan 15")
    static func formatShortDate(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "MMM d"
        return formatter.string(from: date)
    }

    /// Format a time for display (e.g., "2:30 PM")
    static func formatTime(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "h:mm a"
        return formatter.string(from: date)
    }

    /// Format a date with weekday (e.g., "Monday, January 15, 2024")
    static func formatFullDate(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "EEEE, MMMM d, yyyy"
        return formatter.string(from: date)
    }

    /// Get today's date formatted (e.g., "January 15")
    static var todayFormatted: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "MMMM d"
        return formatter.string(from: Date())
    }

    /// Group measurements by date string key
    static func groupByDate<T>(_ items: [T], dateExtractor: (T) -> Date) -> [(date: String, items: [T])] {
        var groups: [String: [T]] = [:]

        for item in items {
            let dateKey = formatDate(dateExtractor(item))
            if groups[dateKey] == nil {
                groups[dateKey] = []
            }
            groups[dateKey]?.append(item)
        }

        return groups.map { (date: $0.key, items: $0.value) }
            .sorted { item1, item2 in
                guard let date1 = items.first(where: { formatDate(dateExtractor($0)) == item1.date }).map(dateExtractor),
                      let date2 = items.first(where: { formatDate(dateExtractor($0)) == item2.date }).map(dateExtractor) else {
                    return false
                }
                return date1 > date2
            }
    }
}
